### PR TITLE
[MOB-2446] - Online vs Offline handling

### PR DESCRIPTION
**Pull request by Ayyanchira**
## 🔹 Jira Ticket(s) if any

https://iterable.atlassian.net/browse/MOB-2446

## ✏️ Description

1. Added hashset of strings for all offline api
2. Passing request to task schedular only if requests are there in hashset
3. added a constructor for testing purpose.
4. Added test method to check if api's are handled correctly

### CODE REVIEW

File: iterableapi/src/main/java/com/iterable/iterableapi/OfflineRequestProcessor.java [hunk 1]
```diff
@@ -5,22 +5,42 @@
...
 
+import java.util.Arrays;
 import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Set;
 
 class OfflineRequestProcessor implements RequestProcessor {
     private TaskScheduler taskScheduler;
     private IterableTaskRunner taskRunner;
+    Set<String> offlineApiSet = new HashSet<>(Arrays.asList(
```
Comment (vbabenkoru) [inline iterableapi/src/main/java/com/iterable/iterableapi/OfflineRequestProcessor.java:21]:
```suggestion
    private static final Set<String> offlineApiSet = new HashSet<>(Arrays.asList(
```

File: iterableapi/src/test/java/com/iterable/iterableapi/OfflineRequestProcessorTest.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    @Test
+    public void testNonOfflineRequestIsNotStored() throws JSONException {
+        IterableApiRequest request = new IterableApiRequest("apiKey", IterableConstants.PUSH_DISABLE_AFTER_REGISTRATION, new JSONObject(), "POST", null, null, null);
+        offlineRequestProcessor.processPostRequest(request.apiKey, request.resourcePath, request.json, request.authToken, request.successCallback, request.failureCallback);
+        verify(mockTaskStorage, times(0)).createTask(IterableConstants.ENDPOINT_TRACK_INAPP_CLICK, IterableTaskType.API, request.toJSONObject().toString());
+    }
+
+    @Test
+    public void testAllOfflineApisUseTaskSchedular() {
```
Comment (vbabenkoru) [inline iterableapi/src/test/java/com/iterable/iterableapi/OfflineRequestProcessorTest.java:45]:
```suggestion
    public void testAllOfflineApisUseTaskScheduler() {
```

File: iterableapi/src/main/java/com/iterable/iterableapi/OfflineRequestProcessor.java [hunk 3]
```diff
@@ -5,22 +5,42 @@
...
     OfflineRequestProcessor(Context context) {
         IterableTaskStorage taskStorage = IterableTaskStorage.sharedInstance(context);
         taskRunner = new IterableTaskRunner(taskStorage, IterableActivityMonitor.getInstance());
         taskScheduler = new TaskScheduler(taskStorage, taskRunner);
     }
 
+    @VisibleForTesting
+    OfflineRequestProcessor(Context context, IterableTaskStorage storage, IterableTaskRunner iterableTaskRunner) {
+        taskRunner = iterableTaskRunner;
+        taskScheduler = new TaskScheduler(storage, taskRunner);
```
Comment (vbabenkoru) [inline iterableapi/src/main/java/com/iterable/iterableapi/OfflineRequestProcessor.java:41]: You can inject just `taskRunner` and `taskScheduler` as dependencies, skipping `storage`.

File: iterableapi/src/test/java/com/iterable/iterableapi/OfflineRequestProcessorTest.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        mockTaskScheduler = mock(TaskScheduler.class);
+        offlineRequestProcessor = new OfflineRequestProcessor(mockTaskScheduler, mockTaskRunner);
+    }
+
+    @Test
+    public void testOfflineRequestIsStored() throws JSONException {
+        IterableApiRequest request = new IterableApiRequest("apiKey", IterableConstants.ENDPOINT_TRACK_INAPP_CLICK, new JSONObject(), "POST", null, new IterableHelper.SuccessHandler() {
+            @Override
+            public void onSuccess(@NonNull JSONObject data) {
+                IterableLogger.v("OfflineRequestProcessorTest", "OnSuccess callback executed");
```
Comment (vbabenkoru) [inline iterableapi/src/test/java/com/iterable/iterableapi/OfflineRequestProcessorTest.java:38]: What are these logging lines for? It seems like they are not part of the test logic.

File: iterableapi/src/test/java/com/iterable/iterableapi/OfflineRequestProcessorTest.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+                IterableLogger.e("OfflineRequestProcessorTest", "OnFailure callback executed");
+            }
+        });
+        offlineRequestProcessor.processPostRequest(request.apiKey, request.resourcePath, request.json, request.authToken, request.successCallback, request.failureCallback);
+        verify(mockTaskScheduler).scheduleTask(any(IterableApiRequest.class), any(IterableHelper.SuccessHandler.class), any(IterableHelper.FailureHandler.class));
+    }
+
+    @Test
+    public void testNonOfflineRequestIsNotStored() throws JSONException {
+        IterableApiRequest request = new IterableApiRequest("apiKey", IterableConstants.PUSH_DISABLE_AFTER_REGISTRATION, new JSONObject(), "POST", null, null, null);
```
Comment (vbabenkoru) [inline iterableapi/src/test/java/com/iterable/iterableapi/OfflineRequestProcessorTest.java:52]: Why `PUSH_DISABLE_AFTER_REGISTRATION`? It's not an API request path.

Comment (Ayyanchira) [reply iterableapi/src/test/java/com/iterable/iterableapi/OfflineRequestProcessorTest.java]: Oh.. I thought its disableDevice... My bad😅

File: iterableapi/src/test/java/com/iterable/iterableapi/OfflineRequestProcessorTest.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        mockTaskRunner = mock(IterableTaskRunner.class);
+        mockTaskScheduler = mock(TaskScheduler.class);
+        offlineRequestProcessor = new OfflineRequestProcessor(mockTaskScheduler, mockTaskRunner);
+    }
+
+    @Test
+    public void testOfflineRequestIsStored() {
+        IterableApiRequest request = new IterableApiRequest("apiKey", IterableConstants.ENDPOINT_TRACK_INAPP_CLICK, new JSONObject(), "POST", null, null, null);
+        offlineRequestProcessor.processPostRequest(request.apiKey, request.resourcePath, request.json, request.authToken, request.successCallback, request.failureCallback);
+        verify(mockTaskScheduler).scheduleTask(any(IterableApiRequest.class), (IterableHelper.SuccessHandler) isNull(), (IterableHelper.FailureHandler) isNull());
```
Comment (vbabenkoru) [inline iterableapi/src/test/java/com/iterable/iterableapi/OfflineRequestProcessorTest.java:34]:
```suggestion
        verify(mockTaskScheduler).scheduleTask(any(IterableApiRequest.class), isNull(IterableHelper.SuccessHandler.class), isNull(IterableHelper.FailureHandler.class));
```

**Review by vbabenkoru — APPROVED** [review]

**Pull request merged**